src/CADL.Extension/Emitter.Csharp/src/emitter.ts

```diff
@@ -0,0 +1,339 @@
+                  )
+                : resolvePath(options.outputFile);
+
+        const root = createModel(program);
+        // await program.host.writeFile(outPath, prettierOutput(JSON.stringify(root, null, 2)));
+        await program.host.writeFile(
+            outPath,
+            prettierOutput(stringifyRefs(root, null, 1, PreserveType.Objects))
+        );
+        const yamlOutPath = resolvePath(
+            options.outputFile?.replace(".json", `.yaml`)
+        );
+        await program.host.writeFile(yamlOutPath, dump(root));
+    }
+}
+
+function prettierOutput(output: string) {
+    return output + "\n";
+}
+function getClient(
+    clients: InputClient[],
+    clientName: string
+): InputClient | undefined {
+    for (const client of clients) {
+        if (client.Name == clientName) return client;
+    }
+
+    return undefined;
+}
+
+function createModel(program: Program): any {
+    const serviceNamespaceType = getServiceNamespace(program);
+    if (!serviceNamespaceType) {
+        return;
+    }
+    const title = getServiceTitle(program);
+    const version = getServiceVersion(program);
+    if (version === "0000-00-00") {
+        console.error("No API-Version provided.");
+        return;
+    }
+    const description = getDoc(program, serviceNamespaceType);
+    const externalDocs = getExternalDocs(program, serviceNamespaceType);
+
+    const servers = getServers(program, serviceNamespaceType);
+
+    const apiVersions: string[] = [];
+    apiVersions.push(version);
+    const namespace =
+        getServiceNamespaceString(program)?.toLowerCase() || "client";
+    try {
+        const [routes] = getAllRoutes(program);
+        console.log("routes:" + routes.length);
+        const clients: InputClient[] = [];
+
+        //create endpoint parameter from servers
+        let endPointParam = undefined;
+        if (servers !== undefined) {
+            const calServers = resolveServers(program, servers);
+            if (calServers.length > 0) {
+                /* choose the first server as endpoint. */
+                endPointParam = calServers[0].parameters[0];
+            }
+        }
+
+        const apiVersionParam: InputParameter = {
+            Name: "apiVersion",
+            NameInRequest: "apiVersion",
+            Description: "",
+            Type: new InputType("string", InputTypeKind.String, false),
+            Location: RequestLocation.Query,
+            IsRequired: true,
+            IsApiVersion: true,
+            IsContentType: false,
+            IsEndpoint: false,
+            IsResourceParameter: false,
+            SkipUrlEncoding: false,
+            Explode: false,
+            Kind: InputOperationParameterKind.Client
+        };
+        for (const operation of routes) {
+            console.log(JSON.stringify(operation.path));
+            const groupName: string = operation.groupName;
+            let client = getClient(clients, groupName);
+            if (client === undefined) {
```

Comment:
I would check null as well. InputClient could be null.

Reply:
Actually, I hope all the places you check the undefined could check null as well to make the code more robust.

Reply:
No. It will return null. If it does not exist, it will return undefined.